### PR TITLE
feat(theming): support font-family css variable

### PR DESCRIPTION
**Pull request by esotrope**
Before, the material font-family could only be set at application build-time. Now, developers can modify the material font-family at runtime by setting the --mat-font-family css variable. This will greatly improve the application's ability to be themed dynamically at runtime and with WYSIWYG material theme editors.

This commit partially addresses this feature request: https://github.com/angular/material2/issues/4352

### CODE REVIEW

File: src/lib/core/typography/_typography.scss [hunk 1]
```diff
@@ -64,7 +65,10 @@
   }
 
   // Add the base font family to the config.
-  @return map-merge($config, (font-family: $font-family));
+  @return map-merge($config, (
+    font-family: $font-family,
+    use-css-variables: $use-css-variables
+  ));
```
Comment (esotrope) [inline src/lib/core/typography/_typography.scss]: Example usage in Angular:
```scss
// styles.scss
@import '~@angular/material/theming';

$typography-config: map-merge(
    mat-typography-config(),
    (use-css-variables: true)
);

@include mat-core($typography-config);
```
```ts
// app.component.ts
@Component({
  selector: 'app-root',
  template: `
    <h1>Font Family CSS Variable Demo</h1>
  `
})
export class AppComponent {

  constructor(private element: ElementRef) {
    this.element.nativeElement.style.setProperty('--mat-font-family', 'Comic Sans MS');
  }

}
```

**Comment by esotrope** [comment]
closing to simplify developer setup using suggested `@mixin` on thread

File: src/lib/core/style/_css-variables.scss [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+}
+
+// Wraps the provided fallback value in a css variable declaration
+//
+// @param $variable The name of the css variable to use (without the "--mat-" prefix)
+// @param $fallback The value to use when the css variable is not set
+//                  or not supported by the browser
+@function css-var($variable, $fallback) {
+  @return var(--mat-#{$variable}, $fallback);
+}
```
Comment (esotrope) [inline src/lib/core/style/_css-variables.scss]: These methods can be reused when expanding css variable support to colors

File: src/lib/core/typography/_typography-utils.scss [hunk 3]
```diff
@@ -34,6 +36,11 @@
   @return if($font-family == null, $font-family, unquote($font-family));
 }
 
+// Creates a CSS rule for font-family using the --mat-font-family css variable
+@mixin mat-font-family-css-var($config) {
+  @include mat-css-var(font-family, font-family, mat-font-family($config));
+}
```
Comment (esotrope) [inline src/lib/core/typography/_typography-utils.scss]: convenience method to wrap mat-font-family calls and provide the necessary css-variable css rules

**Comment by esotrope** [comment]
Demo of setting the --mat-font-family css variable:
change-font-angular-mat

**Comment by jelbourn** [comment]
Thanks for the PR!

We unfortunately cannot use CSS variables until we end support for IE11. Once that happens, we will most likely change the entire theming system to use them. See #7374

**Comment by esotrope** [comment]
Thank you for the quick feedback!

This will be bad news for my client however.
Would people be open to creating a parallel `_theming.scss` (maybe `_css-variable-based-theming.scss`) which people can optionally include instead to mitigate the bundle size issue?

I understand this would not be fun to maintain as it would likely necessitate parallel scss partials for all existing component stylesheets. However, I'm still curious if a PR of that nature would be welcome since i will likely be developing the code regardless of whether or not the main fork adopts it : /

**Comment by jelbourn** [comment]
We unfortunately don't have the resources to maintain two parallel sets of theming approaches. If there were tooling that automatically did this, perhaps, but that would be a pretty significant undertaking.

**Pull request closed without merging**